Draft PR: Working on issue #7824

Arup-Chauhan opened the pull request:
Working on issue #7824

Arup-Chauhan commented:
Hi @daniel-rhoades mentioned to use the [makeHighContrastColor](https://eui.elastic.co/#/theming/colors/utilities#contrast) function in the code.

Also, I did sign the CLA after raising the PR when I saw the check fail, is it okay or will I need a new PR later?

Arup-Chauhan commented:
> @Arup-Chauhan My preference would be to skip `makeHighContrastColor` and custom text color support. To be frank, it's not something I want to support in EuiBadge or EuiAvatar right now.
> 
> I think the simplest path forward would be to DRY out the shared color contrast logic between EuiAvatar and EuiBadge to an exported utility (which we can locate in `src/services/`). This will enable EuiAvatar to also throw a console warning if color contrast is insufficient.

so we are looking at  a shared utility for the color contrast logic and place it in src/services/. This will eliminate duplicate implementations in EuiAvatar and EuiBadge. By using this utility, EuiAvatar can also throw a console warning if the color contrast is insufficient.

Is this a correct way?

cee-chen commented:
That's correct!

Arup-Chauhan commented:
> That's correct!

Cool, thanks for the feedback, I will start doing it and let you know how it goes, have a good day!

Arup-Chauhan commented:
@cee-chen Hi, just wanted to update you that I am in the middle of some summer break commitments from college, will resume working on this soon asap.

Hope that works!

Arup-Chauhan commented:
@cee-chen Hi Cee, wanted to share that I have kind of coded a logic using the constrst utility, will push the changes in this draft repo for your review by the end of this week.

Thanks for the patience, the college semester course load is a little tough this time so spending my major amount there!

cee-chen commented:
No worries at all. This isn't a high priority task, so take your time!

Arup-Chauhan commented:
> No worries at all. This isn't a high priority task, so take your time!


Hi @cee-chen , I hope you're doing well. I noticed that the issue is marked as closed. I wasn’t able to update you earlier due to some urgent academic commitments. If it’s still possible, I would love to continue working on this.

I've made some progress on creating the shared color contrast utility to eliminate duplicate implementations in components like `EuiAvatar` and `EuiBadge`. Below is a snippet showing the approach for calculating the color contrast ratio:

```typescript
/**
 * Calculates the contrast ratio between two colors.
 * @param {string} foreground - The foreground color in hexadecimal format.
 * @param {string} background - The background color in hexadecimal format.
 * @returns {number} - The contrast ratio between the two colors.
 */
function calculateContrast(foreground: string, background: string): number {
    const lum1 = getLuminance(foreground);
    const lum2 = getLuminance(background);
    const brightest = Math.max(lum1, lum2);
    const darkest = Math.min(lum1, lum2);
    return (brightest + 0.05) / (darkest + 0.05);
}
```

This utility will help ensure the colors meet accessibility standards. Let me know if I can proceed with further refinements or if there are any changes you'd like me to incorporate.

Looking forward to your response!

cee-chen commented:
There isn't a need for a custom `calculateContrast` function - we already have this via our `chroma.js` library:

https://github.com/elastic/eui/blob/5c4031565d1f0fef7e08e2377ea5b29bf8963b54/packages/eui/src/components/badge/color_utils.ts#L68-L70

You should be able to simply re-use the existing util and EuiBadge's logic to throw an error when color contrast is too low:

https://github.com/elastic/eui/blob/7fac1cedcc9f9313c66d358b0437d0b52c990ba8/packages/eui/src/components/badge/badge.tsx#L142-L153

My preferred approach would be moving the `getColorContrast` from EuiBadge's color utils to `packages/eui/src/services/color/contrast.ts`, and adding a new `warnIfContrastBelowMin()` util.

Arup-Chauhan commented:
Hi @cee-chen, thanks for being patient as work through on this issue and thanks for your feedback

I have made the following changes and pushed it in the draft PR too.

Replaced the existing contrast-checking logic in `EuiBadge` with calls to the centralized utility functions as suggested.

1. Created the centralized `getColorContrast` and `warnIfContrastBelowMin()` utilities in `packages/eui/src/services/color/contrast.ts`:

```typescript
import chroma from 'chroma-js';

export const getColorContrast = (textColor: string, backgroundColor: string): number => {
  return chroma.contrast(textColor, backgroundColor);
};

export const warnIfContrastBelowMin = (textColor: string, backgroundColor: string, wcagContrastMin: number): void => {
  const contrastRatio = getColorContrast(textColor, backgroundColor);
  if (contrastRatio < wcagContrastMin) {
    console.warn(
      `Warning: ${backgroundColor} background with ${textColor} text has a low contrast ratio of ${contrastRatio.toFixed(
        2
      )}. Should be above ${wcagContrastMin}.`
    );
  }
};
```

2. Replaced the existing logic in `EuiBadge`:

**Original code:**

```typescript
const contrastRatio = getColorContrast(textColor, color);
if (contrastRatio < wcagContrastMin) {
  console.warn(
    `Warning: ${color} badge has a low contrast of ${contrastRatio.toFixed(
      2
    )}. Should be above ${wcagContrastMin}.`
  );
}
```

**Replaced with:**

```typescript
import { warnIfContrastBelowMin } from '../../../services/color/contrast';

warnIfContrastBelowMin(textColor, color, wcagContrastMin);
```

This change centralizes the contrast-checking logic, leveraging `chroma.js` as recommended for improved consistency and maintainability. Please let me know if there are any further adjustments or feedback!



Also, I would like to know how to check the changes and verify them

tkajtoch commented:
Hi @Arup-Chauhan! Thank you for your contribution!

This PR has been open as a draft for a while now. Do you think you'll be able to finish your work soon? If not, we'll likely close this PR, but feel free to create a new one when your changes are ready for review.

Arup-Chauhan commented:
> Hi @Arup-Chauhan! Thank you for your contribution!
> 
> This PR has been open as a draft for a while now. Do you think you'll be able to finish your work soon? If not, we'll likely close this PR, but feel free to create a new one when your changes are ready for review.

Hi @tkajtoch I would love to resume working on this, I made some changes on my way to work on this and hence wanted to some feedback from @cee-chen .

But in case he is occupied, could you please look at my code change that I am planning to do and give me some inputs on it?
Thanks!

Arup-Chauhan commented:
> Hey @Arup-Chauhan 👋🏻 if you mark this PR as "Ready for review", the whole team will be pinged and reminded to check your PR!

Hi @weronikaolejniczak @tkajtoch @cee-chen 


Hope you are well, and wanted to check-in:

Before marking this as ready for review, I wanted to request you to have a look into the code change that I have proposed, if that's a good direction, I can mark this as a ready-to-review.

acstll commented:
@Arup-Chauhan hey, new team member here :)

I went through the code quickly and it's definitely in the right direction 👍 

You could mark it as ready already. If you clean it up, even better.

Thanks again for taking the time 🙏

Arup-Chauhan commented:
@acstll thank you for the feedback!

I will try cleaning up from side, and then will try to push some changes!

weronikaolejniczak commented:
@Arup-Chauhan please take your time! Let us know in case you need some guidance. Let's see this PR through! 😄

Arup-Chauhan commented:
> @Arup-Chauhan please take your time! Let us know in case you need some guidance. Let's see this PR through! 😄

Sure thank you! @weronikaolejniczak
